switools/swisignature.py

```diff
@@ -144,9 +160,123 @@ def signSwiHandler( args ):
    rootCaFile = args.CAfile
    signatureFile = args.signature
    signingKeyFile = args.key
-   signSwi( swi, signingCertFile, rootCaFile, signatureFile, signingKeyFile )
+   signSwiAll( swi, signingCertFile, rootCaFile, signatureFile, signingKeyFile )
    print( 'SWI/X file %s successfully signed and verified.' % swi )
 
+def extractSignature( swi, destFile, sigFileName=None ):
+   sigFn = sigFileName
+   if not sigFn:
+      sigFn = signaturelib.SWI_SIG_FILE_NAME
+   ret = os.system( "set -e;"
+                    "destDir=$(readlink -f $(dirname %s));"
+                    "destFile=$(basename %s);"
+                    "swi=$(readlink -f %s);"
+                    "sigFile=%s;"
+                    "cd $destDir;"
+                    "unzip -o -q $swi $sigFile;"
+                    "mv $sigFile $destFile" % (
+                      destFile, destFile, swi, sigFn )
+                  )
+   if ret != 0:
```

Comment:
Latest code uses subprocess everywhere.
Note that when somebody runs "swi-signature <foo>" at a bash prompt he does not need to craft a malicious <foo> to cause mayhem, he can just run it directly...

Reply:
zipfile.py is very nasty: it requires zip version 2.0 to extract things it inserted...

Reply:
Which is why I only suggested using `zipfile` when extracting files or checking the contents of an existing zip 😃

Reply:
My dislike of `os.system` isn't just its potential for allowing shell injection attacks, it's also that concatenating strings then splitting them again is prone to failure if one of the substrings contains a space e.g. in a pathname, and as there's a perfectly good alternative available (i.e. `subprocess`) then why not use it?

Reply:
Good point about spaces in filenames, maybe there are people that are gluttons for punishment out there.
Well, we can quote the path and the swi in the os.system call if some people ask for troubles and put spaces in filenames...

Anyway, I tried it and it does not work as is:
```
> ./test > t; diff t t2
30c30
< Sand-4GB: SWI/X verification failed.
---
> Sand-4GB: SWI/X verification successful.
33c33
< retCode: 255
---
> retCode: 0

> git diff
diff --git a/switools/swisignature.py b/switools/swisignature.py
index b846412..ab6cd28 100644
--- a/switools/swisignature.py
+++ b/switools/swisignature.py
@@ -179,11 +179,19 @@ def insertSignature( swi, sigFileName, workDir ):
 def extractSignature( swi, destFile ):
    sigFn = signaturelib.SWI_SIG_FILE_NAME
    destDir = os.path.dirname( destFile )
-   cmd = "unzip -o -q %s %s" % ( os.path.abspath( swi ), sigFn )
-   if not runCmd( cmd, destDir ):
+  #cmd = "unzip -o -q %s %s" % ( os.path.abspath( swi ), sigFn )
+  #if not runCmd( cmd, destDir ):
+  #   msg = "Error: Cannot extract signature file %s from  %s" % ( sigFn, swi )
+  #   raise SwiSignException( SWI_SIGN_RESULT.ERROR_SIGNATURE_EXTRACTION_FAILED, msg )
+  #os.rename( "%s/%s" % ( destDir, sigFn ), destFile )
+
+   try:
+      with zipfile.ZipFile( swi ) as zf:
+         zf.extract( sigFn, destDir )
+      os.rename( "%s/%s" % ( destDir, sigFn ), destFile )
+   except Exception:
       msg = "Error: Cannot extract signature file %s from  %s" % ( sigFn, swi )
       raise SwiSignException( SWI_SIGN_RESULT.ERROR_SIGNATURE_EXTRACTION_FAILED, msg )
-   os.rename( "%s/%s" % ( destDir, sigFn ), destFile )

 # get signature from remote signing server
 def getSignatureFile( swi, signatureFile ):
diff --git a/switools/swisigningservice.py b/switools/swisigningservice.py
index 0da549d..04705d2 100755
--- a/switools/swisigningservice.py
+++ b/switools/swisigningservice.py
@@ -1,4 +1,3 @@
-#!/usr/bin/env python3.7
 import os
 import tempfile
 import sys
```

Reply:
The weird part is that only one of the inner signature fails with your code:
```
> /usr/local/bin/verify-swi EOSmini.swi --CAfile /etc/swi-signing-devCA/root.crt
Optimizations in EOSmini.swi: Default Sand-4GB Strata-4GB
Default: SWI/X verification successful.
Sand-4GB: SWI/X verification failed.
Strata-4GB: SWI/X verification successful.
SWI/X verification successful.
retCode: 255
stderr:
```
some flushing issues?

Reply:
> Well, we can quote the path and the swi in the os.system call if some people ask for troubles and put spaces in filenames...
Which is why I think it's better to use `subprocess.check_call` and pass it a list so it doesn't matter if any arguments contain spaces (or any other special characters).

Reply:
> The weird part is that only one of the inner signature fails with your code

I can't see any obvious reason why that change would cause this.  I'll try to reproduce it.

Reply:
I can't reproduce this: it verifies ok every time I run it.  What I do notice though is that all of the sha256 sums are different each time I run it on a SWI.
```
(venv) [johnc@us200 ~/work]$ swi-signature prepare --outfile ./EOS.swi  --force-sign ../EOS.swi
27210b1056d406e09da97b4244a7c4443226fb0d04bceb0bc76d36a4edbade5b
(venv) [johnc@us200 ~/work]$ swi-signature sign EOS.swi /etc/swi-signing-devCA/signing.crt /etc/swi-signing-devCA/root.crt --key /etc/swi-signing-devCA/signing.key
Optimizations in EOS.swi: Default Sand-4GB Strata-4GB
Default sha256: cf9f3e15ca740a54578a6031325e5cdb88122f912565270b86a74f58813afb06
Sand-4GB sha256: 93fba52a4f5761dcc2752c7ee6929ec0925643ace9a30a819f5b77a331897ecb
Strata-4GB sha256: ea8adf67fbc7a6c679475ba7b99bc7a578b18a1748cdf6dec72a05b5685b2bb2
Adding signature files to EOS.swi: Default.signature Sand-4GB.signature Strata-4GB.signature
EOS.swi sha256: 731ab15ef7eb784def9162efc6418dba881e02175aa25d5d7a90d38ab95debd7
SWI/X file EOS.swi successfully signed and verified.
(venv) [johnc@us200 ~/work]$ verify-swi EOS.swi --CAfile /etc/swi-signing-devCA/root.crt
Optimizations in EOS.swi: Default Sand-4GB Strata-4GB
Default: SWI/X verification successful.
Sand-4GB: SWI/X verification successful.
Strata-4GB: SWI/X verification successful.
SWI/X verification successful.
(venv) [johnc@us200 ~/work]$
(venv) [johnc@us200 ~/work]$
(venv) [johnc@us200 ~/work]$ rm EOS.swi
(venv) [johnc@us200 ~/work]$ swi-signature prepare --outfile ./EOS.swi  --force-sign ../EOS.swi
c4f84e794a5d7404f13099f9ce9b25cdfe76c264f47066ee3beb64cc172decab
(venv) [johnc@us200 ~/work]$ swi-signature sign EOS.swi /etc/swi-signing-devCA/signing.crt /etc/swi-signing-devCA/root.crt --key /etc/swi-signing-devCA/signing.key
Optimizations in EOS.swi: Default Sand-4GB Strata-4GB
Default sha256: 219a21b01e0a4b074ea0c7098d3566ad89276c39e7aa879e8f6efc185cacefb7
Sand-4GB sha256: aa438f6d7b96c682d0bcca21195d4805bdc926c5da6992599ed0ada363323001
Strata-4GB sha256: 6d1f64b584e5abb599300c296c0bbeb9fbfd5c7d304f23c990d3d707b2d1858c
Adding signature files to EOS.swi: Default.signature Sand-4GB.signature Strata-4GB.signature
EOS.swi sha256: 5408ae5c0501f10dbf22c39bc5b2cdae705faf9fd46847f7d0eec02d0671b0d2
SWI/X file EOS.swi successfully signed and verified.
(venv) [johnc@us200 ~/work]$ verify-swi EOS.swi --CAfile /etc/swi-signing-devCA/root.crt
Optimizations in EOS.swi: Default Sand-4GB Strata-4GB
Default: SWI/X verification successful.
Sand-4GB: SWI/X verification successful.
Strata-4GB: SWI/X verification successful.
SWI/X verification successful.
```

Reply:
I've pushed a branch with all of my suggested fixes applied to my fork of this repo: https://github.com/johnrclarke/swi-tools.git, branch swim-fixes

Reply:
The new null-signature we insert each time is probably what changes the sha256 each time, because of the timestamp.